src/network/network_connection.cpp

```diff
@@ -615,202 +609,132 @@ void NetworkConnection::Reset() {
+        LOG_ERROR("Unexpected ProcessResult");
+        return Transition::ERROR;
+    }
+  } else {
+    ProcessResult status = protocol_handler_->Process(rbuf_, (size_t) thread_id);
+    switch (status) {
+      case ProcessResult::MORE_DATA_REQUIRED:
+        return Transition::NEED_DATA;
+      case ProcessResult::COMPLETE:
+        return Transition::PROCEED;
+      case ProcessResult::PROCESSING: {
+        if (event_del(network_event) == -1) {
+          //TODO: There may be better way to handle this error
+          LOG_ERROR("Failed to delete event");
           PL_ASSERT(false);
         }
-        conn->protocol_handler_->GetResult();
-        conn->traffic_cop_.SetQueuing(false);
-        conn->TransitState(ConnState::CONN_WRITE);
-        break;
+        LOG_TRACE("ProcessResult: queueing");
+        return Transition::GET_RESULT;
       }
+      case ProcessResult::TERMINATE:
+        return Transition::ERROR;
+    }
+  }
 
-      case ConnState::CONN_WRITE: {
-        // examine write packets result
-        switch (conn->WritePackets()) {
-          case WriteState::WRITE_COMPLETE: {
-            if (!conn->UpdateEvent(EV_READ | EV_PERSIST)) {
-              LOG_ERROR("Failed to update event, closing");
-              conn->TransitState(ConnState::CONN_CLOSING);
-              break;
-            }
-            conn->TransitState(ConnState::CONN_PROCESS);
-            break;
-          }
-
-          case WriteState::WRITE_NOT_READY: {
-            // we can't write right now. Exit state machine
-            // and wait for next callback
-            done = true;
-            break;
-          }
+  // Should not be here
+  return Transition::ERROR;
+}
 
-          case WriteState::WRITE_ERROR: {
-            LOG_ERROR("Error during write, closing connection");
-            conn->TransitState(ConnState::CONN_CLOSING);
-            break;
-          }
-        }
-        break;
+Transition NetworkConnection::ProcessWrite() {
+  // TODO(tianyu): Should convert to use Transition in the top level method
+  switch (WritePackets()) {
+    case WriteState::WRITE_COMPLETE: {
+      if (!UpdateEvent(EV_READ | EV_PERSIST)) {
+        LOG_ERROR("Failed to update event, closing");
+        return Transition::ERROR;
       }
+      return Transition::PROCEED;
+    }
 
-      case ConnState::CONN_CLOSING: {
-        conn->CloseSocket();
-        done = true;
-        break;
-      }
+    case WriteState::WRITE_NOT_READY:
+      return Transition::NONE;
 
-      case ConnState::CONN_CLOSED: {
-        done = true;
-        break;
-      }
+    case WriteState::WRITE_ERROR:
+      LOG_ERROR("Error during write, closing connection");
+      return Transition::ERROR;
+  }
 
-      case ConnState::CONN_INVALID: {
-        PL_ASSERT(false);
-        break;
-      }
-    }
+  // Should not be here
+  return Transition::ERROR;
+}
+
+Transition NetworkConnection::GetResult() {
+  // TODO(tianyu) We probably can collapse this state with some other state.
```

Comment:
WAIT state I suppose.

Reply:
Will do that in a separate PR to clean up all the states